openedx/openedx-platform
# New configuration for third_party_auth & shibboleth/SAML SSO

**Description**
This PR contains three major changes:
1. This PR refactors `third_party_auth` to use [keyed](https://github.com/edx/edx-platform/pull/8077) [ConfigurationModel](http://vengefulpickle.com/posts/config-model.html)s for most configuration. This has the following benefits:
   - Auth provider configuration can be changed at runtime (no need to reboot server, fast and easy to make changes)
   - Auth provider configuration is edited with the Django admin UI, which is much friendlier than editing `lms.(auth|env).json` files.
   - Now administrators can easily configure their instances to use _any_ of the python-social-auth [OAuth2 backends](https://github.com/omab/python-social-auth/tree/master/social/backends), without writing a single line of code. Before, only Google/Facebook/LinkedIn were available, and adding more required writing new Python classes within a Django app.
   - The third_party_auth code is considerably simplified (lots of code removed from settings.py and provider.py)
2. Shibboleth/SAML provider metadata is now fetched, parsed, and stored in the database by a `manage.py` command.
3. Shibboleth/SAML providers can be configured to require an `eduPersonEntitlement`. This can be used, for example, to only allow a third party University's students to authenticate with SSO but not guests or alumni (who still have valid Shibboleth credentials from that University).

This PR is against the [Shibboleth/SAML SSO feature branch, `feature/shibboleth-tpa`](https://github.com/edx/edx-platform/pull/8140).

**Notes**:
- Configuring Google/Facebook/LinkedIn has changed: you must enter the name ("Google"), icon ("fa-google-plus"), key, and secret into the Django admin UI. You also must choose the backend ("google-oauth2") from a dropdown menu.
- When upgrading an existing installation, the data migration will copy the third party auth settings over from ~/lms.auth.json to the new ConfigurationModel.
- All features should work whether the `FEATURES` flag `ENABLE_COMBINED_LOGIN_REGISTRATION` is true or false.

**Coming in the future**:
- I'd like to get this approach reviewed now before I get too carried away writing tests, so new tests will come later, either as a separate PR or part of this PR.
- Set up the `manage.py saml pull` task to run via celery beat

**Screenshots**:

New SAML Provider configuration UI:
![New SAML Provider configuration UI](https://cloud.githubusercontent.com/assets/945577/7775700/31f72fc4-006a-11e5-9761-def50f3b80b6.png)

**Sandbox** (Updated May 26):
There is a demo at http://sandbox5.opencraft.com/admin/third_party_auth/ - user name 'admin', ping Braden via HipChat or IRC or email for the password.

**Reviewers**:
Code: @smarnach, @cpennington 

**Test Instructions**:
Provider Configuration:
1. Go to http://sandbox5.opencraft.com/admin/third_party_auth/oauth2providerconfig/
2. Click "Add"
3. Click "Enabled", set icon to "fa-google-plus", Name to "Google" backend to "google-oauth2", client ID and secret to "test", and hit save. 
4. A "Sign in with Google" button should now be visible at http://sandbox5.opencraft.com/login if opened in an incognito window
5. Back on the admin change list for OAuth provider configuration, click the "Update" link next to the Google entry, then uncheck "Enabled" and save changes. 
6. Refresh the login page seen in an incognito window. The Google button should immediately be gone.

Provider configuration migration:
1. On a devstack or other instance that is currently configured to use Google/Facebook/LinkedIn as a social auth provider, check out this branch and run `paver update_db`. 
2. Go to http://localhost:8000/admin/third_party_auth/oauth2providerconfig/
3. The old provider[s] configuration should now be seen in the new admin screen
4. Run the LMS and verify that third party auth via that provider[s] is working

SAML Configuration:
1. (If on a local instance only - skip this step for the sandbox): Go to /third_party_auth/samlconfiguration/ and click "Add SAML Configuration". Check "Enabled". Under the "Private Key" field is an openssl command. Run that command, then paste the contents of the resulting "saml.key" into "Private Key" and "saml.crt" into "Public Key". Set "Entity ID" to a unique URL like "http://my-devstack.example.com". Edit "Organization Info" if desired.
2. Go to [/auth/saml/metadata.xml](http://sandbox5.opencraft.com/auth/saml/metadata.xml) and verify that the metadata is working. Save this file to your desktop.
3. Go to https://www.testshib.org/register.html and upload the metadata file from step 2.
4. Go to /admin/third_party_auth/samlproviderconfig/ 
5. Click "Add Provider Configuration". 
6. Check "Enabled", set "Icon Class" to "fa-university", set "Name" to "TestShib [something]" (e.g. "TestShib 1"), set "Backend" to the only option available, set "IdP Slug" to "testshib-1"
7. Set "Entity ID" to "https://idp.testshib.org/idp/shibboleth" and "Metadata source" to "https://www.testshib.org/metadata/testshib-providers.xml"
8. Set "Email attribute" to "urn:oid:1.3.6.1.4.1.5923.1.1.1.6"
9. Save
10. (If on a local instance): Run the command `./manage.py lms saml pull --settings=devstack` and verify that it worked by checking /admin/third_party_auth/samlproviderdata/
11. Open the LMS in an incognito window and try logging in using the new "TestShib 1" button.
12. Edit the new SAML provider configuration and change the entity ID to something different/invalid.
13. Go to the login page and click the "TestShib 1" button. Verify that a friendly error message is seen. (Note: friendly error will only be seen on the sandbox or if DEBUG is False).
